aesgcm.go

```diff
@@ -0,0 +1,83 @@
+package main
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/rand"
+	"encoding/hex"
+	"strings"
+
+	"golang.org/x/crypto/scrypt"
+)
+
+func DeriveKey(password string, salt []byte) ([]byte, []byte, error) {
+	if salt == nil {
+		salt = make([]byte, 32)
+		if _, err := rand.Read(salt); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	key, err := scrypt.Key([]byte(password), salt, 131072, 8, 1, 32)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return key, salt, nil
+}
+
+func AesGcmEncrypt(plaintext string, password string) (string, error) {
+	secretKey, salt, err := DeriveKey(password, nil)
+	if err != nil {
+		return "", err
+	}
+	plaintextBytes := []byte(plaintext)
+
+	aes, err := aes.NewCipher([]byte(secretKey))
+	if err != nil {
+		return "", err
+	}
+
+	aesgcm, err := cipher.NewGCM(aes)
+	if err != nil {
+		return "", err
+	}
+
+	nonce := make([]byte, aesgcm.NonceSize())
+	_, err = rand.Read(nonce)
+	if err != nil {
+		return "", err
+	}
+
+	ciphertext := aesgcm.Seal(nil, nonce, plaintextBytes, nil)
+
+	return hex.EncodeToString(salt) + "-" + hex.EncodeToString(nonce) + "-" + hex.EncodeToString(ciphertext), nil
+}
+
+func AesGcmDecrypt(ciphertext string, password string) (string, error) {
+	arr := strings.Split(ciphertext, "-")
```

Comment:
This function doesn't throw an error if the incorrect password is used. Is there anything we can do here?

Reply:
we have to, I don't know yet how